packages/runtime/container-runtime/src/runningSummarizer.ts

```diff
@@ -309,38 +321,99 @@ export class RunningSummarizer implements IDisposable {
         })().finally(() => {
             summarizingLock.resolve();
             this.summarizingLock = undefined;
-            this.checkRerunHeuristics();
+            this.checkSummarizeAgain();
         }).catch((error) => {
             this.logger.sendErrorEvent({ eventName: "UnexpectedSummarizeError" }, error);
         });
     }
 
     public summarizeOnDemand(
         reason: string,
-        options: Omit<ISubmitSummaryOptions, "summaryLogger">,
+        { afterSequenceNumber, ...options }: IOnDemandSummarizeOptions,
     ): OnDemandSummarizeResult {
-        // Check for concurrent summary attempts. If one is found,
-        // return a promise that caller can await before trying again.
-        if (this.summarizingLock !== undefined) {
-            // The heuristics are blocking concurrent summarize attempts.
-            return { alreadyRunning: this.summarizingLock };
+        if (this.stopping || this.disposed) {
+            const failBuilder = new SummarizeResultBuilder();
+            failBuilder.fail("RunningSummarizer stopped or disposed", undefined);
+            return failBuilder.build();
         }
-        if (this.generator.isSummarizing()) {
-            // Another summary is currently being generated.
-            return { alreadyRunning: this.generator.waitSummarizing() };
+        const onDemandReason = `onDemand;${reason}` as const;
+        if (afterSequenceNumber === undefined) {
+            // Check for concurrent summary attempts. If one is found,
+            // return a promise that caller can await before trying again.
+            if (this.summarizingLock !== undefined) {
+                // The heuristics are blocking concurrent summarize attempts.
+                return { alreadyRunning: this.summarizingLock };
+            }
+            if (this.generator.isSummarizing()) {
+                // Another summary is currently being generated.
+                return { alreadyRunning: this.generator.waitSummarizing() };
+            }
+            const result = this.generator.summarize(onDemandReason, options);
+            result.receivedSummaryAckOrNack.finally(() => this.checkSummarizeAgain());
+            return result;
+        }
+        // If afterSequenceNumber is provided, the on-demand summary is "enqueued" to run
+        // once an eligible op comes in with sequenceNumber >= afterSequenceNumber.
+        if (this.enqueuedSummary !== undefined) {
+            // Note that it could be already generating a summary at this point in time.
+            // In that case, the result of this is misleading for earlier sequence numbers.
+            // We could instead fail the previously enqueued summary, but would require a cancel.
+            this.enqueuedSummary.reason = onDemandReason;
+            this.enqueuedSummary.afterSequenceNumber = afterSequenceNumber;
+            this.enqueuedSummary.options = options;
+        } else {
+            this.enqueuedSummary = {
```

Comment:
What happens if op with sequenceNumber = afterSequenceNumber has already been received? We should summarize right away if that's the case right.

Reply:
Yeah, good call, I'll add a call to tryRunEnqueuedSummary at the end.